recipe-client-addon/lib/EventEmitter.jsm

```diff
@@ -1,34 +1,42 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-
-// This file is meant to run inside action sandboxes
-
 "use strict";
 
+const {utils: Cu} = Components;
+const {jsonCopy} = Cu.import("resource://shield-recipe-client/lib/Utils.jsm", {});
+Cu.import("resource://shield-recipe-client/lib/LogManager.jsm");
+
+this.EXPORTED_SYMBOLS = ["EventEmitter"];
 
-this.EventEmitter = function(driver) {
-  if (!driver) {
-    throw new Error("driver must be provided");
-  }
+const log = LogManager.getLogger("event-emitter");
 
+this.EventEmitter = function(sandboxManager) {
   const listeners = {};
 
   return {
+    createSandboxedEmitter() {
+      return sandboxManager.cloneInto({
+        on: this.on.bind(this),
+        off: this.off.bind(this),
+        once: this.once.bind(this),
+      });
+    },
+
     emit(eventName, event) {
       // Fire events async
       Promise.resolve()
         .then(() => {
           if (!(eventName in listeners)) {
-            driver.log(`EventEmitter: Event fired with no listeners: ${eventName}`);
+            log.info(`EventEmitter: Event fired with no listeners: ${eventName}`);
             return;
           }
-          // freeze event to prevent handlers from modifying it
-          const frozenEvent = Object.freeze(event);
           // Clone callbacks array to avoid problems with mutation while iterating
           const callbacks = Array.from(listeners[eventName]);
           for (const cb of callbacks) {
-            cb(frozenEvent);
+            // Copy event to prevent modifcations leaking between callbacks.
+            const copiedEvent = jsonCopy(event);
+            cb(sandboxManager.cloneInto(copiedEvent));
```

Comment:
Oh, so actually, the `cloneInto` here will guarantee every callback gets a fresh copy anyway.